File: python/dungeon.py
from math import floor
import random
import copy


random.seed()
MAPSIZE = 32
global pairId 
pairId = 0

class Leaf:
	def __init__(self, x, y, xend, yend, direction, depth, id):
		self.x = x
		self.y = y
		self.xend = xend
		self.yend = yend
		self.splitDirection = direction
		self.depth = depth
		self.pairId = id
	def xSize(self):
		return self.xend - self.x
	def ySize(self):
		return self.yend - self.y
	
class Corridor:
	def __init__(self, x, y, xend, yend):
		self.x = x
		self.y = y
		self.xend = xend
		self.yend = yend

def clamp(value, min, max):
	if (value < min):
		return min
	if (value > max):
		return max
	return value

def shrink(leaf):

	if (leaf.xSize() > 5):
		xmid = floor((leaf.xend + leaf.x) / 2)
		x = random.randint(leaf.x, xmid)
		xend = random.randint(xmid, leaf.xend)
		leaf.x = x
		leaf.xend = xend
	if (leaf.ySize() > 5):
		ymid = floor((leaf.yend + leaf.y) / 2)
		y = random.randint(leaf.y, ymid)
		yend = random.randint(ymid, leaf.yend)
		leaf.y = y
		leaf.yend = yend




def divide(leaf, depth, horizontal=True):
	global pairId
	pairId += 1
	x = leaf.x
	y = leaf.y
	xend = leaf.xend
	yend = leaf.yend

	if (leaf.xSize() > 5 and leaf.ySize() > 5):
		xmid = random.randint(x+2, xend -2)
		ymid = random.randint(y+2, yend -2)

		if (horizontal ):
			leaf1 = Leaf(leaf.x, leaf.y, xmid-1, leaf.yend, "h", depth, pairId)
			leaf2 = Leaf(xmid+1, leaf.y, leaf.xend, leaf.yend, "h", depth, pairId)
			return [leaf1, leaf2]
		elif ((not horizontal)):
			leaf1 = Leaf(leaf.x, leaf.y, leaf.xend, ymid-1, "v", depth, pairId)
			leaf2 = Leaf(leaf.x, ymid+1, leaf.xend, leaf.yend, "v", depth, pairId)
			return [leaf1, leaf2]
		else:
			return [leaf]
	else:
		return [leaf]


def generate(leaf, maxDepth, depth = 1):
	direction = random.choice([True, False])
	leaves = divide(leaf, depth, direction)
	roomLeaves = []
	leafTree = copy.deepcopy(leaves)
	if (depth < maxDepth):
		for dividedLeaf in leaves:
			newLeaves, newCompTree = generate(dividedLeaf, maxDepth, depth + 1)
			roomLeaves += newLeaves
			leafTree += newCompTree
	else:
		roomLeaves = leaves

	return roomLeaves, leafTree



map = [["##" for _ in range(MAPSIZE)] for _ in range(MAPSIZE)]

rootLeaf = Leaf(1,1,MAPSIZE-2, MAPSIZE-2, 0,  None, 0)

leaves, compTree = generate(rootLeaf, 5)
corridors = []


for i in range(len(compTree)):
	first = compTree[i]
	for j in range(len(compTree)):
		second = compTree[j]
		if (second.pairId == first.pairId and first != second):
			firstMidX = floor((first.x + first.xend) / 2)
			firstMidY = floor((first.y + first.yend) / 2)
			secondMidX = floor((second.x + second.xend) / 2)
			secondMidY = floor((second.y + second.yend) / 2)
			x = firstMidX
			y = firstMidY
			endx = secondMidX
			endy = secondMidY
			"""
			if (firstMidX > secondMidX):
				x = secondMidX
				endx = firstMidX
			if (firstMidY > secondMidY):
				y = secondMidY
				endy = firstMidY
			"""
			corridor = Corridor(x, y, endx, endy)
			corridors.append(corridor)
			break


for corridor in corridors:
	for y in range(corridor.y, corridor.yend+1):
		for x in range (corridor.x, corridor.xend+1):
			map[y][x] = ".."
			pass

for leaf in leaves:
	shrink(leaf)
	pass


for leaf in leaves:
	#if (leaf.depth == 3):
	for y in range(leaf.y, leaf.yend+1):
		for x in range (leaf.x, leaf.xend+1):
			map[y][x] = ".."
			pass

for y in range(MAPSIZE):
	for x in range(MAPSIZE):
		print(map[y][x], sep="", end="")
	print("")

File: include/dungeon.c
#include "dungeon.h"
#include <tonc.h>

static int PAIRID = 0;
static int ID = 0;
static int MAXITERATIONS = 100;
static int MINROOMSIZE = 5;

struct Leaf {
	int active;
	int x;
	int y;
	int xend;
	int yend;
	int pairId;
	int id;
};



void divide(struct Leaf* leaves, struct Leaf* leaf, int horizontal) {
	PAIRID++;

	int xSize = leaf->xend - leaf->x;
	int ySize = leaf->yend - leaf->y;

	int xmid, ymid;
	if (xSize > MINROOMSIZE && ySize > MINROOMSIZE && ID < MAXITERATIONS) {
		xmid = qran_range(leaf->x+2, leaf->xend -1);
		ymid = qran_range(leaf->y+2, leaf->yend -1);

		if (horizontal) {

			leaves[0].active = 1;
			leaves[0].x = leaf->x;
			leaves[0].y = leaf->y;
			leaves[0].xend = xmid - 1;
			leaves[0].yend = leaf->yend;
			leaves[0].pairId = PAIRID;
			leaves[0].id = ID; 
			ID++;

			leaves[1].active = 1;
			leaves[1].x = xmid + 1;
			leaves[1].y = leaf->y;
			leaves[1].xend = leaf->xend;
			leaves[1].yend = leaf->yend;
			leaves[1].pairId = PAIRID;
			leaves[1].id = ID;
			ID++;

		}
		else { //vertical

			leaves[0].active = 1;
			leaves[0].x = leaf->x;
			leaves[0].y = leaf->y;
			leaves[0].xend = leaf->xend;
			leaves[0].yend = ymid - 1;
			leaves[0].pairId = PAIRID;
			leaves[0].id = ID; 
			ID++;


			leaves[1].active = 1;
			leaves[1].x = leaf->x;
			leaves[1].y = ymid + 1;
			leaves[1].xend = leaf->xend;
			leaves[1].yend = leaf->yend;
			leaves[1].pairId = PAIRID;
			leaves[1].id = ID; 
			ID++;


		}

	}
	else {
		
		leaves[0].active = 1;
		leaves[0].x = leaf->x;
		leaves[0].y = leaf->y;
		leaves[0].xend = leaf->xend;
		leaves[0].yend = leaf->yend;
		leaves[0].pairId = leaf->pairId;
		leaves[0].id = ID; 
		ID++;

	
	}
}

void generate(struct Leaf* leaf, struct Leaf* finalLeaves, struct Leaf* pairTree, int maxDepth, int depth) {

	int direction = qran_range(0, 2);
	struct Leaf leaves[2];
	leaves[0].active = 0;
	leaves[1].active = 0;
	divide(&leaves, leaf, direction);

	for (int i = 0; i < MAXITERATIONS; i++) {
		if (!pairTree[i].active) {
			if (leaves[0].active && leaves[1].active) {
				//printf("OK\n");
                int x = 0;
                int y = 0;
                int xend = 0;
                int yend = 0;

                getCorridor(&leaves[0], &leaves[1], &x, &y, &xend, &yend);
                pairTree[i].x = x;
                pairTree[i].y = y;
                pairTree[i].xend = xend;
                pairTree[i].yend = yend;
                pairTree[i].active = 1;

			}
			break;
		}
	}


	if (depth < maxDepth) {
		for (int i = 0; i < 2; i++) {
			if (leaves[i].active) {

				generate(&leaves[i], finalLeaves, pairTree, maxDepth, depth+1);
			}
		}
	}
	else {
		for (int i = 0; i < MAXITERATIONS; i++) {
			if (!finalLeaves[i].active) {
				finalLeaves[i] = leaves[0];
				finalLeaves[i+1] = leaves[1];
				break;
			}
		}
		
	}

}

void getCorridor(struct Leaf* first, struct Leaf* second, int* x, int* y, int* xend, int* yend) {

	int firstMidX = (first->x + first->xend) / 2;
	int secondMidX = (second->x + second->xend) / 2;
	int firstMidY = (first->y + first->yend) / 2;
	int secondMidY = (second->y + second->yend) / 2;

	*x = firstMidX;
	*y = firstMidY;
	*xend = secondMidX;
	*yend = secondMidY;

	if (*x > *xend) {
		int temp = *x;
		*x = *xend;
		*xend = temp;
	}

	if (*y > *yend) {
		int temp = *y;
		*y = *yend;
		*yend = temp;
	}
}


void shrink(struct Leaf* leaves) {
    for (int i = 0; i < MAXITERATIONS; i++) {
        struct Leaf* leaf = &leaves[i];
        if (!leaf->active) {
            continue;
        }
        else {
            int xSize = leaf->xend - leaf->x;
            int ySize = leaf->yend - leaf->y;
            if (xSize > 5) {
                int xmid = (leaf->xend + leaf->x) / 2;
                int x = qran_range(leaf->x, xmid+1);
        		int xend = qran_range(xmid, leaf->xend+1);
        		leaf->x = x;
        		leaf->xend = xend;
            }
            if (ySize > 5) {
                int ymid = (leaf->yend + leaf->y) / 2;
                int y = qran_range(leaf->y, ymid+1);
        		int yend = qran_range(ymid, leaf->yend);
        		leaf->y = y;
        		leaf->yend = yend;
            }
        }
    }
}


void getDungeon(char* map, int mapsize, FIXED* playerX, FIXED* playerY) {
	PAIRID = 0;
	ID = 0;
	struct Leaf rootLeaf;

	rootLeaf.x = 1;
	rootLeaf.y = 1;
	rootLeaf.xend = mapsize-1;
	rootLeaf.yend = mapsize-1;

	struct Leaf finalLeaves[MAXITERATIONS];
	struct Leaf pairTree[MAXITERATIONS];

	//initialize the array of leaves that are unused
	for (int i = 0; i < MAXITERATIONS; i++) {
		finalLeaves[i].active = 0;
		pairTree[i].active = 0;
	}

	generate(&rootLeaf, &finalLeaves, &pairTree, 6, 0);
	shrink(&finalLeaves);
	
	for (int y = 0; y < mapsize; y++) {
		for (int x = 0; x < mapsize; x++) {
			map[mapsize*y+x] = 1;
		}
	}

	for (int i = 0; i < MAXITERATIONS; i++) {
		struct Leaf iter = finalLeaves[i];
		if (!iter.active) {
			continue;
		}

		for (int y = 0; y < mapsize-1; y++) {
			for (int x = 0; x < mapsize-1; x++) {
				if (x >= iter.x && y >= iter.y && x <= iter.xend && y <= iter.yend) {
					map[mapsize*y+x] = 0;
				}
			}
		}
	}

	for (int i = 0; i < MAXITERATIONS; i++) {
		struct Leaf first = pairTree[i];
		if (!first.active) {
			continue;
		}
		                				

        for (int x = first.x; x < first.xend+1; x++) {
	    	for (int y = first.y; y < first.yend+1; y++) {
	    		map[mapsize*y+x] = 0;
	    	}
	    }
	}	

	//no longer needed, as player is placed relative to the door
	
	for (int x = 0; x < mapsize; x++) {
		for (int y = 0; y < mapsize; y++) {
			if (map[mapsize*y+x] == 0) {
				*playerX = int2fx(x) + 128;
				*playerY = int2fx(y) + 128;
				return;
			}
		}
	}
	

}

